### ISSUE TITLE
Color - Sender Onboarding 

### ISSUE THREAD

**Issue by sliu1000** [CDCgov/prime-reportstream#6550]
## Problem statement

We received an onboarding request from sender (Christina Sheehan - publichealthreporting@color.com) at the end of July.  The would like to move forward and onboard as a sender to RS.  

## What you need to know

- They have completed the TOS
- Have some sample test files created (has not sent us yet)
- Still needs to have a Kickoff meeting/call.

## Acceptance criteria

-Confirm with user on a kickoff meeting/email for initiation.
-obtain the test files for review

## TO DO
- [x] Send email to Color to initiate Kickoff via email
- [ ] Create individual tickets for this epic for each of the onboarding steps (BX can work on doing this)

**Comment by BerniXiongA6** [comment]
@sarahnw to decide whether we Kickoff via email or live meeting. 
Last email from Color:

> Hi Sharon,
> Thank you for your response. The following dates would work for us: 
>  Wednesday, August 31st, 9:00 - 10:00 AM PT
>  Friday, September 2nd, 11:00 AM to 1:00 PM PT
>  Thursday, September 8th, any time after 9:00 AM PT
> Alternatively, Katie mentioned the possibility of an async form of kick-off via email. If that's an acceptable or quicker option, we are happy to do so whenever is convenient for your engineering team. 
> 
> Best,
> Christina

cc: @sliu1000

**Comment by BerniXiongA6** [comment]
@BerniXiongA6 sent out initial Kickoff info via email with a cc: to RS shared inbox and @sarahnw -- see below:

> Hello Color Team!
> 
> My name is Berni Xiong, and I'm the Delivery Manager from the ReportStream team that onboards new users. We would be happy to complete our ReportTeam Sender Kickoff process with you all via email.
> 
> 
> 
> To acclimate you to this process, I'm including some useful onboarding resources in the following. Please note that for today's email, we'll cover a brief introduction to ReportStream (see bullet # 1 below) and we'll also outline our onboarding steps for data senders (see bullet # 1 and 2 below).
> 
> 
> 
> 
> API Programmer's Guide: This guide provides full documentation for interacting with the ReportStream API. When reviewing the document, please start with the Introduction (page 4) and Onboarding Process sections (page 6-8), before reviewing the rest of the document just to get you acquainted with ReportStream before you take a deeper dive into the onboarding steps. (Link: https://reportstream.cdc.gov/resources/programmers-guide)
> 
> Getting Started and Onboarding Steps: This slide below outlines how the onboarding process works for ReportStream data senders. These steps are detailed in the programmer's guide linked above; however, we wanted you to have an at-a-glance view of the onboarding steps.
> 
> 
>  
> 
> TIMING: When you are ready to proceed, the first step of "Onboarding to ReportStream" is to send sample data (start with page 6 of the programmer's guide above). At this time, we've got about a 2-week backlog before we can look at the first sample file, so you have plenty of time to get one prepared. Until then, please let us know if you have any questions!
> 
>  
> 
> Thanks,
> Berni

**Comment by BerniXiongA6** [comment]
@MauriceReeves-usds If I'm missing other tickets, please feel free to create them under this epic. At this time, I'll create a handful to match the steps we outline in the API programmer's guide.

**Comment by shy2cdc** [comment]
@MauriceReeves-usds @BerniXiongA6 

Hi Bernadette and Maurice,

Hope you had a good weekend. Thank you so much for the introduction. I look forward to working with you all on this project. 

Whenever you are ready, we have prepared initial non-PII test messages. Could you please let us know how to best send them to you - would email work? 

Christina

**Comment by BerniXiongA6** [comment]
Hi @shy2cdc @sliu1000 we created an [individual ticket #6599](https://app.zenhub.com/workspaces/onboarding--operations-6166edbd409257001e09f1b2/issues/cdcgov/prime-reportstream/6599) for @MauriceReeves-usds to work with the sender.

**Comment by sliu1000** [comment]
Update from Sender:

I wanted to check in with you on the test HL7 messages we sent over last week. Have you had a chance to review them and, if so, do you have any feedback for our team?

**Comment by sliu1000** [comment]
Update from Color:

The team working on the ReportStream project had a few general questions about the program and the API integration that they were not able to answer with the programmer's guide. Would there be a time the teams could meet next week or could we send you a short list of questions via email if a meeting wouldn't be possible?

Has an engineer been assigned to this?

**Comment by sliu1000** [comment]
Followup Question from Color:

Hi @sarahnw ,

Happy Friday! We wanted to quickly check-in on if there's anything on our end that could help this along or if you have a sense of how long it takes to generate a client for us? We are waiting to begin development after we receive the client. 

Please let us know if it would be possible for us to meet next week! We're very excited to be adopting ReportStream (it's incredibly valuable for reportable diseases). 

Ruey
On Thu, Sep 22, 2022 at 4:35 PM Ruey Pham <ruey.pham@color.com> wrote:
Thank you for the quick response! 


We have generated public & private keys today and you can find the public key attached. Hope we are able to begin sending dummy requests to ReportStream soon. 


Follow-up Questions: 


We’ve attached a PEM file containing our public key for the staging environment. Is this format okay for you? 


We’re assuming we should generate a separate keypair for stage and production. Is this correct?


What client ID should we use in the client_credentials request? 


Please let us know when you have an update about availability to schedule a Q&A meeting.


Ruey



On Wed, Sep 21, 2022 at 2:30 PM White, Sarah (CDC/DDPHSS/OD/HITSSU) <tlk9@cdc.gov> wrote:
Hi Ruey,

Answers to the best of my ability in line:
Can you give us an idea of how many messages are being processed each day? And how many folks have adopted ReportStream?
ReportStream has been adopted by in 30+ jurisdictions and many, many more senders. We have no traffic issues.
Can we generate a public / private key pair for JWT-based auth (and you'll store the public key on your end) ? Otherwise, if we need to use the shared API key method, can you generate this and share it with us in Keybase?
We can handle either method, though token-based is preferred.
When are you supposed to use individual vs asynchronous APIs? Is there a threshold? 
This is up to sender preference. There's a 10k record limit in a single submission.
Are there plans to add other infectious diseases to ReportStream in the future? If this is a possibility, how would we hear about new diseases?
Yes there are! We're actively working to expand beyond COVID-19. When we have more concrete information to share we'll post about it on our website (reportstream.cdc.gov) and email our current users. We haven't set up an email list people can subscribe to yet, but if we do we'll let you know.
Let us know if there's anything else we can answer,
Sarah

**Comment by sliu1000** [comment]
Updated Correspondence:

9/30/2022

Hi Berni,

Thank you for your response and for providing the ticket number. Monday sounds great for a progress update—we look forward to hearing from you all.

Have a great weekend!

Best,
Christina 


9/30/2022
Hi Christina, thanks for following up. 

Since our team is already supporting this work we’ve previously started with you, there’s no need to submit a new ticket. I’m out this afternoon and Maurice is out all day today. Can we follow up with you on Monday with progress updates?

Also, your ticket number is #6599.

Thanks,
Berni

9/30/2022

Hi Bernadette and Maurice,

Happy Friday! We received the email about the new support ticketing system your team is using - congratulations on the launch! 
Have you had a chance to review our test HL7 messages or could you assist us in creating an account for the support site so we can resubmit the test messages in a ticket for review? 

Thank you for your help.

Best,
Christina

**Comment by sliu1000** [comment]
Update from Sender:

Hi Team,


Thank you so much for taking the time to chat with us today. Meeting notes can be found in our meeting doc. A few action items & questions from our meeting below: 


ReportStream


Sent message and received "errors" : [ {"scope" : "parameter","message" : "'client:color-labs': unknown schema 'hl7/color-labs-covid-19'"}]


ReportStream to complete Color’s setup so Color can begin testing in staging - question: do you know when folks might be able to finish setting up? We’d love to continue the process to integrate with ReportStream and are excited to adopt the product.


For confirmation - which fields in the payload are used to distribute the result to the corresponding state DPH?


Color


Will generate and send PROD keys to ReportStream


Thanks!


Ruey

**Comment by sliu1000** [comment]
Update from Color:

Hi Sharon, 

Thanks for confirming receipt. I'm also attaching our production public key, another item from Ruey's meeting notes. We'd love it if someone could get this added to our client in prod in tandem with the investigation of the staging messaging issue highlighted above. This would greatly help us in our testing. 

Best, 

Nick

**Comment by BerniXiongA6** [comment]
@jbiskie can help handoff to new engineer if not done before leaving project cc: @BerniXiongA6 @MauriceReeves-usds

**Comment by sliu1000** [comment]
Update from Color sent on 10/27/2022:

Hi Team,

We're currently working on our eng design for ReportStream and was wondering if you guys might have availability from Wed - Fri (inclusive) next week to meet to go over a few questions that we have? Availability in PT below:

Wed, Nov 2 - 9:30am-10:30am, 12-2:15pm, after 3pm
Thurs, Nov 3 - after 10am
Fri, Nov 4 - after 12pm

Please let me know if you guys have time/prefer another time.

Thanks!

Ruey

**Comment by BerniXiongA6** [comment]
BX emailed Color on 10/31/2022:

> Hi Color Team, 
> 
> We're glad you were able to connect to the staging environment and will begin testing API. It sounds as though you have some questions for ReportStream and we're happy to meet with you on Thursday and answer any questions you have. I've copied James Gilmore, our HL7 expert who will be handling your onboarding moving forward. 
> 
> Thanks,
> Berni Xiong

**Comment by BerniXiongA6** [comment]
Scheduled to meet w/ Color on Thurs. Nov. 3rd 3pm ET -- @BerniXiongA6 and @GilmoreA6 cc: @brandonnava @sliu1000

**Comment by shy2cdc** [comment]
Email from Ruey:

Hi!

We're currently working on the ReportStream integration and had a few questions come up (doc version here if it's easier to share and distribute).

Questions: 

Patient address state and collection facility state is how the logic distro the reports. What are the exact fields used for routing?


When reporting results, the response body contains “id” and “submission_id”. One is a UUID and the other an integer.


Are both these values guaranteed to be unique with each submission/request?


Is there a functional distinction between them?


Which is the preferred value to query the history API with?


We had asked about the partial batch success feature before and it sounded like it was on by default.  It’ll be much more efficient for us if this feature is turned off.


Can we turn off partial batch message success?


Is it possible to turn this off in the staging environment relatively soon, so it can be tested?


Some questions regarding your logic for routing records to destinations.


Could you describe this logic generally?


If it’s difficult to describe fully, could you provide us with the key fields being used to make these determinations?


What happens to a record that should be routed to a state that doesn’t have a ReportStream integration? Is there a default destination or is it simply rejected or does something else happen?



Please let us know whenever you get an opportunity or if it's easier for us to meet synchronously.


Thanks!
